From ac3521017602f5c678c4f5fe9405827f87356fff Mon Sep 17 00:00:00 2001
From: Rekha Eswaran <rekha.eswaran@intel.com>
Date: Fri, 26 Jun 2020 14:51:39 +0800
Subject: [PATCH] DRV_DPM_SW-306 - DPM queue information enhancement 	-
 DP_SUBIF_SPECIFIC_QUEUE flag case validity during dp_register 	- update queue
 table information for drop Q

---
 drivers/net/datapath/dpm/datapath_api.c            | 19 ----------
 drivers/net/datapath/dpm/gswip31/datapath_misc.c   | 42 +++++-----------------
 drivers/net/datapath/dpm/gswip31/datapath_ppv4.c   |  7 ++++
 .../net/datapath/dpm/gswip31/datapath_ppv4_api.c   |  3 +-
 4 files changed, 16 insertions(+), 55 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index 990fa47a29eb..1d37d8400e3e 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -806,8 +806,6 @@ int32_t dp_deregister_subif_private(int inst, struct module *owner,
 	struct dp_subif_info *sif;
 	u32 dma_ch_offset, dma_ch_ref;
 	struct inst_info *dp_info = get_dp_prop_info(inst);
-	struct dp_node_alloc port_node;
-	struct q_info *q_info;
 
 	port_id = subif_id->port_id;
 	port_info = get_dp_port_info(inst, port_id);
@@ -898,23 +896,6 @@ int32_t dp_deregister_subif_private(int inst, struct module *owner,
 		if (get_dp_deqport_info(inst, cqm_port)->ref_cnt)
 			continue;
 
-		/* delete all queues which may created by PPA or other apps */
-		port_node.inst = inst;
-		port_node.dp_port = port_id;
-		port_node.type = DP_NODE_PORT;
-		port_node.id.cqm_deq_port = cqm_port;
-
-		if (sif->data_flag_ops & DP_SUBIF_SPECIFIC_Q) {
-			q_info = get_dp_q_info(inst, sif->qid_list[j]);
-
-			if (q_info->flag && !q_info->need_free) {
-				/* Ref Cnt=0 Reset the used flag */
-				if (!q_info->ref_cnt)
-					q_info->flag = 0;
-			}
-		} else {
-			dp_node_children_free(&port_node, 0);
-		}
 		/*disable cqm port */
 		cbm_data.dp_inst = inst;
 		cbm_data.cbm_inst = dp_port_prop[inst].cbm_inst;
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index 6d5a6333dbc1..240173498cd2 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -1570,13 +1570,15 @@ static int free_queue_info(int inst, int dp_port, int qid)
 	struct dp_node_alloc node;
 
 	q_info = get_dp_q_info(inst, qid);
+	if (!q_info->need_free)
+		return DP_FAILURE;
 	if (!q_info->ref_cnt) {
 		pr_err("Why dp_q_tbl[%d][%d].ref_cnt Zero:expect > 0\n",
 		       inst, qid);
 		return DP_FAILURE;
 	}
 	q_info->ref_cnt--;
-	if (q_info->flag && !q_info->ref_cnt && q_info->need_free) {
+	if (q_info->flag && !q_info->ref_cnt) {
 		DP_DEBUG(DP_DBG_FLAG_REG, "Free qid %d\n", qid);
 		node.id.q_id = qid;
 		/*if no subif using this queue, need to delete it*/
@@ -1702,8 +1704,7 @@ static int deq_update_info(struct dp_subif_upd_info *info)
 		/* retrieve the qid corresponding to old deq port */
 		old_def_qid = old_deq_pinfo->first_qid;
 		old_deq_pinfo->ref_cnt--;
-		if (old_def_qid > 0 &&
-		    !free_queue_info(info->inst, info->dp_port, old_def_qid)) {
+		if (!free_queue_info(info->inst, info->dp_port, old_def_qid)) {
 			if (old_deq_pinfo->f_first_qid &&
 			    old_deq_pinfo->first_qid == old_def_qid) {
 				old_deq_pinfo->f_first_qid = 0;
@@ -1899,7 +1900,6 @@ static int subif_hw_set(int inst, int portid, int subif_ix,
 			/* update scheduler table later */
 
 		} else if (q_flag == DP_SUBIF_SPECIFIC_Q) {
-			/* Q is created by Caller and QiD is in Subif Data */
 
 			/* Single GEM or 1 Subif have only 1 Qid
 			 * Multiple GEM or Multiple Subif can have same Qid
@@ -1907,11 +1907,7 @@ static int subif_hw_set(int inst, int portid, int subif_ix,
 			q_info = get_dp_q_info(inst, data->subif_data->q_id);
 
 			if (!q_info->flag) {
-				/* Mark this Q as Used */
-				q_info->flag = 1;
-				q_info->need_free = 0;
-				q_info->ref_cnt = 1;
-
+				q_info->need_free = 0; /* caller Q,so no free */
 				q_info->cqm_dequeue_port = q_port.cqe_deq;
 				deq_pinfo->qos_port = -1;
 				deq_pinfo->ref_cnt++;
@@ -1919,11 +1915,10 @@ static int subif_hw_set(int inst, int portid, int subif_ix,
 					atomic_inc(&(dp_dma_chan_tbl[inst] +
 						     dma_ch_offset)->ref_cnt);
 			} else {
-				/* For Multiple GEM having same Q, we increment
-				 * Q reference count
+				/* Multiple subif can have same Q, we increment
+				 * Q reference count if Queue created by DPM
 				 */
 				q_info->ref_cnt++;
-				get_dp_q_info(inst, q_port.cqe_deq)->ref_cnt++;
 				deq_pinfo->ref_cnt++;
 				if (port_info->num_dma_chan)
 					atomic_inc(&(dp_dma_chan_tbl[inst] +
@@ -2029,7 +2024,6 @@ static int subif_hw_reset(int inst, int portid, int subif_ix,
 	int cqm_deq_port;
 	int dma_ch_offset;
 	struct pmac_port_info *port_info = get_dp_port_info(inst, portid);
-	struct dp_node_alloc node;
 	struct dp_subif_info *sif = get_dp_port_subif(port_info, subif_ix);
 	int bp = sif->bp;
 	struct cqm_port_info *deq_pinfo;
@@ -2050,11 +2044,6 @@ static int subif_hw_reset(int inst, int portid, int subif_ix,
 		q_info = get_dp_q_info(inst, qid);
 
 		/* santity check table */
-		if (!q_info->ref_cnt) {
-			pr_err("Why dp_q_tbl[%d][%d].ref_cnt Zero:expect > 0\n",
-			       inst, qid);
-			return DP_FAILURE;
-		}
 		if (!deq_pinfo->ref_cnt) {
 			pr_err("Why dp_deq_port_tbl[%d][%d].ref_cnt Zero\n",
 			       inst, cqm_deq_port);
@@ -2068,7 +2057,6 @@ static int subif_hw_reset(int inst, int portid, int subif_ix,
 		}
 
 		/* update queue/port/sched/bp_pmapper table's ref_cnt */
-		q_info->ref_cnt--;
 		deq_pinfo->ref_cnt--;
 		if (port_info->num_dma_chan)
 			atomic_dec(&(dp_dma_chan_tbl[inst] +
@@ -2115,18 +2103,7 @@ static int subif_hw_reset(int inst, int portid, int subif_ix,
 
 		deq_pinfo = get_dp_deqport_info(inst, cqm_deq_port);
 
-		if (q_info->flag && !q_info->ref_cnt && q_info->need_free) {
-			DP_DEBUG(DP_DBG_FLAG_QOS, "Free qid %d\n", qid);
-			node.id.q_id = qid;
-			/*if no subif using this queue, need to delete it*/
-			node.inst = inst;
-			node.dp_port = portid;
-			node.type = DP_NODE_QUEUE;
-			dp_node_free(&node, 0);
-
-			/*update dp_q_tbl*/
-			q_info->flag = 0;
-			q_info->need_free = 0;
+		if (!free_queue_info(inst, portid, qid)) {
 			if (deq_pinfo->f_first_qid &&
 			    deq_pinfo->first_qid == qid) {
 				deq_pinfo->f_first_qid = 0;
@@ -2137,9 +2114,6 @@ static int subif_hw_reset(int inst, int portid, int subif_ix,
 					 "dp_deq_port_tbl[%d][%d].f_first_qid reset\n",
 					 inst, cqm_deq_port);
 			}
-		} else {
-			DP_DEBUG(DP_DBG_FLAG_QOS, "q_id[%d] dont need freed\n",
-				 qid);
 		}
 		DP_DEBUG(DP_DBG_FLAG_QOS,
 			 "%s:%s=%d %s=%d q[%d].cnt=%d cqm_p[%d].cnt=%d %s=%d\n",
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c
index fdd4f1fd53ec..eb9ec414f583 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4.c
@@ -549,6 +549,13 @@ int init_ppv4_qos(int inst, int flag)
 		 priv->ppv4_drop_q, q,
 		 priv->cqm_drop_p,
 		 priv->ppv4_drop_p);
+	/* update drop queue/port state in global table */
+	priv->qos_sch_stat[q].parent.node_id = priv->ppv4_drop_p;
+	priv->qos_sch_stat[q].parent.type = DP_NODE_PORT;
+	priv->qos_sch_stat[q].parent.flag = PP_NODE_ACTIVE;
+	priv->qos_sch_stat[priv->ppv4_drop_p].child[0].flag = PP_NODE_ACTIVE;
+	priv->qos_sch_stat[priv->ppv4_drop_p].child[0].node_id = q;
+	priv->qos_sch_stat[priv->ppv4_drop_p].child[0].type = DP_NODE_QUEUE;
 #endif /* end of CONFIG_INTEL_DATAPATH_QOS_HAL */
 	DP_LOCK_INIT(&priv->qos_lock);
 	DP_DEBUG(DP_DBG_FLAG_DBG, "%s done\n", __func__);
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index 8c0169c838b9..a1cb649a4bea 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -3883,8 +3883,7 @@ int dp_queue_map_set_31(struct dp_queue_map_set *cfg, int flag)
 		return DP_FAILURE;
 	}
 
-	if (!is_qid_valid(cfg->q_id) ||
-	    (cfg->q_id != priv->ppv4_drop_q && is_q_node_free(priv, cfg->q_id)))
+	if (!is_qid_valid(cfg->q_id))
 		return DP_FAILURE;
 
 	qmap_cfg.mpe1 = cfg->map.mpe1;
